File: typescript/src/erc8183/quoteVerify.ts
/** Reference verification for ERC-8183 provider quote signatures. */

import type { PublicClient } from "viem";
import { getAddress, hashMessage, parseAbi, recoverMessageAddress } from "viem";
import { canonicalJson, keccakOfText } from "../core/canonicalJson.js";
import { buildDescriptionContent } from "./negotiation.js";

const ERC1271_MAGIC_VALUE = "0x1626ba7e";
const ERC1271_ABI = parseAbi([
  "function isValidSignature(bytes32 hash, bytes signature) view returns (bytes4)",
]);
const HASH_HEX = /^0x[0-9a-fA-F]{64}$/;
const SIGNATURE_HEX = /^0x(?:[0-9a-fA-F]{2})+$/;

export type QuoteSigVerdict =
  | {
      valid: true;
      method: "eip191" | "erc1271";
      signer: `0x${string}`;
    }
  | { valid: false; reason: string };

export interface VerifyQuoteSignatureOpts {
  /** Full `NegotiationResult` envelope or flattened on-chain description. */
  envelope: Record<string, unknown>;
  /** Expected provider account (`job.provider`). */
  provider: `0x${string}`;
  /** Chain reads used for account signatures and time/chain anchoring. */
  publicClient: PublicClient;
  /** Expected commerce/checker contract, when the caller has one in context. */
  expectedVerifyingContract?: `0x${string}`;
  /**
   * Historical acceptance block (normally the `JobFunded` block). Omit
   * before acceptance to verify against current chain state.
   */
  blockNumber?: bigint;
}

function signedContent(
  envelope: Record<string, unknown>,
): Record<string, unknown> {
  if (
    envelope.response !== null &&
    typeof envelope.response === "object" &&
    !Array.isArray(envelope.response)
  ) {
    return buildDescriptionContent(
      envelope,
      (envelope.chain_id as number | undefined) ?? null,
      (envelope.verifying_contract as string | undefined) ?? null,
    );
  }

  return Object.fromEntries(
    Object.entries(envelope).filter(
      ([key]) => key !== "negotiation_hash" && key !== "provider_sig",
    ),
  );
}

function quoteExpiresAt(envelope: Record<string, unknown>): unknown {
  if (
    envelope.response !== null &&
    typeof envelope.response === "object" &&
    !Array.isArray(envelope.response)
  ) {
    return (
      envelope.quote_expires_at ||
      (envelope.response as Record<string, unknown>).quote_expires_at
    );
  }
  return envelope.quote_expires_at;
}

/**
 * Verify that `provider_sig` authenticates the canonical quote content.
 *
 * EOA quotes recover directly from the EIP-191 message. Contract-account
 * quotes fall through to ERC-1271 (implemented below this EOA fast path).
 * Malformed or cryptographically invalid input returns `{ valid: false }`;
 * RPC failures throw because "unavailable" is not a cryptographic verdict.
 *
 * Passing `blockNumber` checks both quote expiry and ERC-1271 authorization
 * at that historical block. This preserves the acceptance-time verdict when
 * a session/checker is revoked later. The RPC must serve historical state.
 * An `eth_call` at a block observes end-of-block state, so exact ordering of
 * acceptance and revocation within one block requires atomic verification by
 * the accepting contract rather than this off-chain helper.
 */
export async function verifyQuoteSignature(
  opts: VerifyQuoteSignatureOpts,
): Promise<QuoteSigVerdict> {
  const negotiationHash = opts.envelope.negotiation_hash;
  const providerSig = opts.envelope.provider_sig;
  if (typeof negotiationHash !== "string" || !HASH_HEX.test(negotiationHash)) {
    return { valid: false, reason: "missing or invalid negotiation_hash" };
  }
  if (typeof providerSig !== "string" || !SIGNATURE_HEX.test(providerSig)) {
    return { valid: false, reason: "missing or invalid provider_sig" };
  }

  const signedVerifyingContract = opts.envelope.verifying_contract;
  let checker: `0x${string}` | undefined;
  if (
    signedVerifyingContract !== undefined &&
    signedVerifyingContract !== null
  ) {
    if (typeof signedVerifyingContract !== "string") {
      return { valid: false, reason: "invalid verifying_contract" };
    }
    try {
      checker = getAddress(signedVerifyingContract);
    } catch {
      return { valid: false, reason: "invalid verifying_contract" };
    }
  }
  if (opts.expectedVerifyingContract) {
    if (!checker) {
      return {
        valid: false,
        reason: "quote is not bound to a verifying_contract",
      };
    }
    if (checker !== getAddress(opts.expectedVerifyingContract)) {
      return { valid: false, reason: "verifying_contract mismatch" };
    }
  }

  const signedChainId = opts.envelope.chain_id;
  if (signedChainId !== undefined && signedChainId !== null) {
    if (
      typeof signedChainId !== "number" ||
      !Number.isSafeInteger(signedChainId) ||
      signedChainId <= 0
    ) {
      return { valid: false, reason: "invalid chain_id" };
    }
    if ((await opts.publicClient.getChainId()) !== signedChainId) {
      return { valid: false, reason: "chain_id mismatch" };
    }
  }

  let recomputed: string;
  try {
    recomputed = keccakOfText(canonicalJson(signedContent(opts.envelope)));
  } catch {
    return { valid: false, reason: "invalid quote content" };
  }
  if (recomputed.toLowerCase() !== negotiationHash.toLowerCase()) {
    return { valid: false, reason: "negotiation_hash mismatch" };
  }

  const expiry = quoteExpiresAt(opts.envelope);
  if (expiry !== undefined && expiry !== null) {
    if (typeof expiry !== "number" || !Number.isSafeInteger(expiry)) {
      return { valid: false, reason: "invalid quote_expires_at" };
    }
    const verificationBlock =
      opts.blockNumber === undefined
        ? await opts.publicClient.getBlock()
        : await opts.publicClient.getBlock({ blockNumber: opts.blockNumber });
    if (BigInt(expiry) <= verificationBlock.timestamp) {
      return { valid: false, reason: "quote has expired" };
    }
  }

  const provider = getAddress(opts.provider);
  try {
    const recovered = await recoverMessageAddress({
      message: negotiationHash,
      signature: providerSig as `0x${string}`,
    });
    if (recovered === provider) {
      return { valid: true, method: "eip191", signer: recovered };
    }
  } catch {
    // ERC-1271 envelopes are not 65-byte EOA signatures; try the account path.
  }

  const block =
    opts.blockNumber === undefined ? {} : { blockNumber: opts.blockNumber };
  const bytecode = await opts.publicClient.getBytecode({
    address: provider,
    ...block,
  });
  if (!bytecode || bytecode === "0x") {
    return { valid: false, reason: "provider signature is not valid" };
  }

  const result = await opts.publicClient.readContract({
    address: provider,
    abi: ERC1271_ABI,
    functionName: "isValidSignature",
    args: [hashMessage(negotiationHash), providerSig as `0x${string}`],
    ...(checker ? { account: checker } : {}),
    ...block,
  });
  if (String(result).toLowerCase() !== ERC1271_MAGIC_VALUE) {
    return { valid: false, reason: "ERC-1271 account rejected provider_sig" };
  }

  return { valid: true, method: "erc1271", signer: provider };
}

File: python/bnbagent/erc8183/quote_verify.py
"""Reference verification for ERC-8183 provider quote signatures."""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any

from eth_account import Account
from eth_account.messages import defunct_hash_message, encode_defunct
from web3 import Web3

from .negotiation import _build_description_content

_ERC1271_MAGIC_VALUE = "0x1626ba7e"
_ERC1271_ABI = [
    {
        "inputs": [
            {"internalType": "bytes32", "name": "hash", "type": "bytes32"},
            {"internalType": "bytes", "name": "signature", "type": "bytes"},
        ],
        "name": "isValidSignature",
        "outputs": [{"internalType": "bytes4", "name": "magicValue", "type": "bytes4"}],
        "stateMutability": "view",
        "type": "function",
    }
]
_HASH_HEX = re.compile(r"^0x[0-9a-fA-F]{64}$")
_SIGNATURE_HEX = re.compile(r"^0x(?:[0-9a-fA-F]{2})+$")


@dataclass(frozen=True)
class QuoteSignatureVerdict:
    """Cryptographic quote-verification result.

    RPC failures deliberately propagate instead of being reported as an
    invalid signature: unavailable historical state is not a cryptographic
    verdict and callers should treat it as retryable.
    """

    valid: bool
    method: str | None = None
    signer: str | None = None
    reason: str | None = None


def _invalid(reason: str) -> QuoteSignatureVerdict:
    return QuoteSignatureVerdict(valid=False, reason=reason)


def _signed_content(envelope: dict[str, Any]) -> dict[str, Any]:
    response = envelope.get("response")
    if isinstance(response, dict):
        return _build_description_content(
            envelope,
            chain_id=envelope.get("chain_id"),
            verifying_contract=envelope.get("verifying_contract"),
        )
    return {
        key: value
        for key, value in envelope.items()
        if key not in {"negotiation_hash", "provider_sig"}
    }


def _quote_expiry(envelope: dict[str, Any]) -> Any:
    response = envelope.get("response")
    if isinstance(response, dict):
        return envelope.get("quote_expires_at") or response.get("quote_expires_at")
    return envelope.get("quote_expires_at")


def verify_quote_signature(
    *,
    envelope: dict[str, Any],
    provider: str,
    w3: Web3,
    expected_verifying_contract: str | None = None,
    block_number: int | None = None,
) -> QuoteSignatureVerdict:
    """Verify an EIP-191 or ERC-1271 provider quote at a chain block.

    ``block_number`` should be the job's ``JobFunded`` block. This anchors
    expiry and contract-account authorization to the moment the buyer
    economically accepted the quote, rather than today's mutable state.
    """

    negotiation_hash = envelope.get("negotiation_hash")
    provider_sig = envelope.get("provider_sig")
    if not isinstance(negotiation_hash, str) or not _HASH_HEX.fullmatch(negotiation_hash):
        return _invalid("missing or invalid negotiation_hash")
    if not isinstance(provider_sig, str) or not _SIGNATURE_HEX.fullmatch(provider_sig):
        return _invalid("missing or invalid provider_sig")

    signed_verifier = envelope.get("verifying_contract")
    checker: str | None = None
    if signed_verifier is not None:
        if not isinstance(signed_verifier, str):
            return _invalid("invalid verifying_contract")
        try:
            checker = Web3.to_checksum_address(signed_verifier)
        except (TypeError, ValueError):
            return _invalid("invalid verifying_contract")
    if expected_verifying_contract is not None:
        if checker is None:
            return _invalid("quote is not bound to a verifying_contract")
        try:
            expected_checker = Web3.to_checksum_address(expected_verifying_contract)
        except (TypeError, ValueError):
            return _invalid("invalid expected verifying_contract")
        if checker != expected_checker:
            return _invalid("verifying_contract mismatch")

    signed_chain_id = envelope.get("chain_id")
    if signed_chain_id is not None:
        if (
            not isinstance(signed_chain_id, int)
            or isinstance(signed_chain_id, bool)
            or signed_chain_id <= 0
        ):
            return _invalid("invalid chain_id")
        if w3.eth.chain_id != signed_chain_id:
            return _invalid("chain_id mismatch")

    try:
        canonical = json.dumps(
            _signed_content(envelope), sort_keys=True, separators=(",", ":")
        )
        recomputed = Web3.keccak(text=canonical).hex()
        if not recomputed.startswith("0x"):
            recomputed = "0x" + recomputed
    except (TypeError, ValueError):
        return _invalid("invalid quote content")
    if recomputed.lower() != negotiation_hash.lower():
        return _invalid("negotiation_hash mismatch")

    expiry = _quote_expiry(envelope)
    if expiry is not None:
        if not isinstance(expiry, int) or isinstance(expiry, bool):
            return _invalid("invalid quote_expires_at")
        block = w3.eth.get_block(block_number if block_number is not None else "latest")
        if expiry <= int(block["timestamp"]):
            return _invalid("quote has expired")

    try:
        provider_address = Web3.to_checksum_address(provider)
    except (TypeError, ValueError):
        return _invalid("invalid provider address")

    try:
        recovered = Account.recover_message(
            encode_defunct(text=negotiation_hash), signature=provider_sig
        )
        if Web3.to_checksum_address(recovered) == provider_address:
            return QuoteSignatureVerdict(
                valid=True, method="eip191", signer=provider_address
            )
    except (TypeError, ValueError):
        # Contract-account signatures are often not 65-byte EOA signatures.
        pass

    block_identifier: int | str = block_number if block_number is not None else "latest"
    bytecode = w3.eth.get_code(provider_address, block_identifier=block_identifier)
    if not bytecode:
        return _invalid("provider signature is not valid")

    contract = w3.eth.contract(address=provider_address, abi=_ERC1271_ABI)
    call_transaction = {"from": checker} if checker else {}
    digest = defunct_hash_message(text=negotiation_hash)
    signature_bytes = bytes.fromhex(provider_sig[2:])
    result = contract.functions.isValidSignature(digest, signature_bytes).call(
        call_transaction,
        block_identifier=block_identifier,
    )
    magic = Web3.to_hex(result).lower()
    if magic != _ERC1271_MAGIC_VALUE:
        return _invalid("ERC-1271 account rejected provider_sig")

    return QuoteSignatureVerdict(valid=True, method="erc1271", signer=provider_address)
